pkg/crd/crd.go

```diff
@@ -0,0 +1,136 @@
+			return fmt.Errorf("failed waiting on CRD '%s': %w", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
+	storedCRDs, err := crdClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list crds: %w", err)
+	}
+
+	// convert existingCRDs to a map for faster lookup
+	existingCRDs := map[string]*apiextv1.CustomResourceDefinition{}
+	for i := range storedCRDs.Items {
+		crd := &storedCRDs.Items[i]
+		existingCRDs[crd.Name] = crd
+	}
+	return existingCRDs, nil
+}
+
+// ensureCRD checks if there is an existing CRD that matches if not it will attempt to create or update the CRD.
+func ensureCRD(ctx context.Context, crd *apiextv1.CustomResourceDefinition, crdClient clientv1.CustomResourceDefinitionInterface, existingCRDs map[string]*apiextv1.CustomResourceDefinition) error {
+	existingCRD, exist := existingCRDs[crd.Name]
+	if !exist {
+		logrus.Infof("Creating embedded CRD %s", crd.Name)
+		_, err := crdClient.Create(ctx, crd, metav1.CreateOptions{})
+		if err == nil {
+			return nil
+		}
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create crd '%s': %w", crd.Name, err)
+		}
+		// item was not in initial list but does exist so we will attempt to get the latest resourceVersion
+		existingCRD, err = crdClient.Get(ctx, crd.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get exiting '%s' CRD: %w", crd.Name, err)
+		}
+		// fallthrough and attempt and update
+	}
+	// only keep the resource version for the desired object
+	crd.ResourceVersion = existingCRD.ResourceVersion
+	// if the CRD exist attempt to update it
+	logrus.Infof("Updating embedded CRD %s", crd.Name)
+	_, err := crdClient.Update(ctx, crd, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update crd '%s': %w", crd.Name, err)
+	}
+	return nil
+}
+
+// waitCRD repeatably checks CRD status until it is established.
+func waitCRD(ctx context.Context, crdName string, crdClient clientv1.CustomResourceDefinitionInterface) error {
+	logrus.Infof("Waiting for CRD %s to become available", crdName)
+	defer logrus.Infof("Done waiting for CRD %s to become available", crdName)
+
+	return wait.PollImmediate(waitInterval, waitDuration, func() (bool, error) {
+		crd, err := crdClient.Get(ctx, crdName, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to get CRD '%s' for status checking: %w", crdName, err)
+		}
+		if crdIsReady(crd) {
+			return true, nil
+		}
+
+		return false, nil
+	})
+}
+
+func crdIsReady(crd *apiextv1.CustomResourceDefinition) bool {
+	for i := range crd.Status.Conditions {
+		cond := &crd.Status.Conditions[i]
+		switch cond.Type {
+		case apiextv1.Established:
+			if cond.Status == apiextv1.ConditionTrue {
+				return true
```

Comment:
Just want to make sure this is intentional. If there are any `Established` status conditions that succeeded, we want to return true? No matter how many errors there may be?

Reply:
Yes, the `condition.Type` is a unique key for all of the conditions in the list.
You may be curious why we don't return false if the condition is not true. This is because we also want to log if the CRD has a naming conflict for the case below, so we have to loop through the entire list to check for both.